### PR TITLE
Book api

### PR DESCRIPTION
**Pull request by mrkiencom**

### CODE REVIEW

File: src/author/author.service.ts [hunk 1]
```diff
@@ -14,6 +16,7 @@ import { AuthorDto } from './dto/author-update.dto';
 export class AuthorService {
   constructor(
     @InjectRepository(AuthorRepository)
+    @Inject(forwardRef(() => AuthorService))
```
Comment (tranmanhlinh) [inline src/author/author.service.ts]: Why you use this decorator?

File: src/author/author.service.ts [hunk 2]
```diff
@@ -31,32 +34,35 @@ export class AuthorService {
     const { name } = updateDto;
     const found = await this.getAuthorById(id);
     if (found) {
-      console.log(found);
-      return await this.authorRepository.save({
-        id: id,
-        name: name,
-      });
+      found.name = name;
+      return await this.authorRepository.save(found);
```
Comment (tranmanhlinh) [inline src/author/author.service.ts]: if (!found) {
  throw E;
}

return await this.authorRepository.save({...found, name })

File: src/author/author.service.ts [hunk 3]
```diff
@@ -31,32 +34,35 @@ export class AuthorService {
...
     } else {
-      await this.authorRepository.delete({ id });
-      return 'delete succes ! ';
+      found.isDelete = true;
+      try {
+        await this.authorRepository.save(found);
+        return 'delete succes ! ';
+      } catch (error) {
+        console.log(error);
+      }
```
Comment (tranmanhlinh) [inline src/author/author.service.ts]: const found = await this.authorRepository.findOne({ where: {id, isDeleted: false}});

if (!found) { throw E; }

return await this.authorRepository.save(found);

// Don't forget to add try catch

File: src/author/author.service.ts [hunk 4]
```diff
@@ -31,32 +34,35 @@ export class AuthorService {
...
     }
   }
   async createAuthor(name: string): Promise<Author> {
     const newAuthor = {
       name: name,
+      isDelete: false,
```
Comment (tranmanhlinh) [inline src/author/author.service.ts]: Set default value of isDeleted is false in the Entity.

File: src/author/author.service.ts [hunk 5]
```diff
@@ -31,32 +34,35 @@ export class AuthorService {
...
     };
     try {
       return await this.authorRepository.save(newAuthor);
     } catch (error) {
       console.log(error);
       if (error.code === 23505) {
-        throw new ConflictException('Author a already exits');
+        throw new ConflictException(Message.ALREADY_EXIST.author);
```
Comment (tranmanhlinh) [inline src/author/author.service.ts]: if (a) { throw A }
throw B.

Remove unnecessary else block.

File: src/book/book.controller.ts [hunk 6]
```diff
@@ -0,0 +1,44 @@
...
+import { BookService } from './book.service';
+import { BookDto } from './dto/book-dto';
+
+@Controller('book')
+@UseGuards(AuthGuard())
+export class BookController {
+  constructor(private bookService: BookService) {}
+  @Get()
+  getBooks(@Query('search') searchBook: string): Promise<Book[]> {
+    return this.bookService.getBooks(searchBook);
```
Comment (tranmanhlinh) [inline src/book/book.controller.ts]: Use @Query as a DTO because we will have may query parameters in the future.

File: src/book/book.entity.ts [hunk 7]
```diff
@@ -0,0 +1,49 @@
...
+import { Column, Entity, ManyToOne, PrimaryGeneratedColumn } from 'typeorm';
+import { BookService } from './book.service';
+
+@Entity()
+export class Book {
+  @PrimaryGeneratedColumn('uuid')
+  id: string;
+  @Column()
+  @IsNotEmpty()
+  title: string;
```
Comment (tranmanhlinh) [inline src/book/book.entity.ts:19]: Use 'class-validator' only for DTO.

File: src/book/book.entity.ts [hunk 8]
```diff
@@ -0,0 +1,49 @@
...
+  @Column({ name: 'author_id' })
+  @IsNotEmpty()
+  authorId: string;
+  @Column({ name: 'category_id' })
+  @IsNotEmpty()
+  categoryId: string;
```
Comment (tranmanhlinh) [inline src/book/book.entity.ts:25]: Add empty lines between every fields.

File: src/book/book.repository.ts [hunk 9]
```diff
@@ -0,0 +1,23 @@
+import { EntityRepository, Repository } from 'typeorm';
+import { Book } from './book.entity';
+
+@EntityRepository(Book)
+export class BookRepository extends Repository<Book> {
+  async getBook(search: string): Promise<Book[]> {
+    if (search) {
+      const query = this.createQueryBuilder('book');
+
+      // query.where('book.title LIKE :search', { search });
```
Comment (tranmanhlinh) [inline src/book/book.repository.ts]: remove comment.

File: src/book/book.repository.ts [hunk 10]
```diff
@@ -0,0 +1,23 @@
...
+      query.andWhere('LOWER(book.title) LIKE LOWER(:search)', {
+        search: `%${search}%`,
+      });
+      const books = await query.getMany();
+      return books.filter((x) => x.isDelete === false);
```
Comment (tranmanhlinh) [inline src/book/book.repository.ts]: Use `isDeleted` instead of `isDelete`

File: src/book/book.service.ts [hunk 11]
```diff
@@ -0,0 +1,144 @@
...
+      publishYear,
+      price,
+      description,
+      cover,
+    } = newBook;
+    try {
+      const categoryFound = await categoryIsExist(categoryId);
+      const authorFound = await authorIsExist(authorId);
+      if (!categoryFound)
+        throw new NotFoundException(Message.NOT_EXIST.cateogory);
```
Comment (tranmanhlinh) [inline src/book/book.service.ts]: The logic should be like this.
- The body of the request will have categoryName and authorName
- The backend will check if there are category or author exist with that name
- If they are exist, save book with their relations
- If they are not exist create a new one with the name.

File: src/book/book.service.ts [hunk 12]
```diff
@@ -0,0 +1,144 @@
...
+        categoryId,
+        publishYear,
+        price,
+        description,
+        cover,
+      } = updateInfoBook;
+
+      try {
+        const categoryFound = await categoryIsExist(categoryId);
+        const authorFound = await authorIsExist(authorId);
```
Comment (tranmanhlinh) [inline src/book/book.service.ts]: Category or author should be optional.

File: src/book/book.service.ts [hunk 13]
```diff
@@ -0,0 +1,144 @@
...
+
+        if (!categoryFound)
+          throw new NotFoundException(Message.NOT_EXIST.cateogory);
+        if (!authorFound) throw new NotFoundException(Message.NOT_EXIST.author);
```
Comment (tranmanhlinh) [inline src/book/book.service.ts]: Please use `if` with a curly block.

File: src/book/book.service.ts [hunk 14]
```diff
@@ -0,0 +1,144 @@
...
+        throw new InternalServerErrorException();
+      }
+    }
+  }
+  async deleteBook(id: string): Promise<string> {
+    const found = this.getBookById(id);
+    if (!found) {
+      throw new NotFoundException(`${Message.NOT_FOUND.book}=${id}`);
+    } else {
+      try {
```
Comment (tranmanhlinh) [inline src/book/book.service.ts]: Check my comment above

File: src/book/book.service.ts [hunk 15]
```diff
@@ -0,0 +1,144 @@
...
+        await this.bookRepository.save({
+          ...found,
+          isDelete: true,
+        });
+        return `delete book with Id=${id} is succes !`;
```
Comment (tranmanhlinh) [inline src/book/book.service.ts]: Return the BookEntity instead

File: src/book/dto/book-dto.ts [hunk 16]
```diff
@@ -0,0 +1,17 @@
...
+  @IsNotEmpty()
+  authorId: string;
+  @IsNotEmpty()
+  categoryId: string;
+  @IsNotEmpty()
+  publishYear: string;
+  @IsNotEmpty()
+  price: string;
+  description: string;
+  @IsNotEmpty()
```
Comment (tranmanhlinh) [inline src/book/dto/book-dto.ts:23]: Check my comment above.

File: src/book/func/check-exist.ts [hunk 17]
```diff
@@ -0,0 +1,29 @@
...
+  }
+}
+export async function authorIsExist(authorId: string): Promise<Author> {
+  try {
+    var found = await getConnection()
+      .getRepository(Author)
+      .createQueryBuilder('author')
+      .where('author.id = :authorId', { authorId })
+      .getOne();
+    return found;
```
Comment (tranmanhlinh) [inline src/book/func/check-exist.ts:25]: const found = authorRepository.fineOne({id:  authorId});

File: src/book/func/get-date.ts [hunk 18]
```diff
@@ -0,0 +1,9 @@
+export const getDateNow = () => {
+  var today = new Date();
+  var dd = String(today.getDate()).padStart(2, '0');
+  var mm = String(today.getMonth() + 1).padStart(2, '0'); //January is 0!
+  var yyyy = today.getFullYear();
```
Comment (tranmanhlinh) [inline src/book/func/get-date.ts:5]: Use dayjs library.

File: src/category/category.service.ts [hunk 19]
```diff
@@ -52,7 +62,7 @@ export class CategoryService {
     if (!found) {
       throw new NotFoundException(`${Message.NOT_FOUND.cateogory}=${id}`);
     } else {
-      await this.categoryRepository.delete({ id });
+      await this.categoryRepository.save({ ...found, isDelete: true });
```
Comment (tranmanhlinh) [inline src/category/category.service.ts]: Return the entity that was deleted or return nothing.

File: src/error/message-eror.ts [hunk 20]
```diff
@@ -1,8 +1,21 @@
...
+    cateogory: 'Category not exist',
+    author: 'Author not exist',
+    book: 'Book not exist',
+    user: 'User not exist',
+  },
+  ALREADY_EXIST: {
+    book: 'Book is already exits',
+    author: 'Author is already exits',
+    category: 'Category is already exits',
+    user: 'User is already exits',
```
Comment (tranmanhlinh) [inline src/error/message-eror.ts]: `exist`

File: src/error/message-eror.ts [hunk 21]
```diff
@@ -1,8 +1,21 @@
 export const Message = {
   CHECK_LOGIN: 'please check your loggin credentials',
   NOT_FOUND: {
-    cateogory: 'Not found Category with ID  ',
-    author: 'Not found Author with ID  ',
-    book: 'Not found Book with ID  ',
+    cateogory: 'Not found Category with ID',
+    author: 'Not found Author with ID ',
```
Comment (tranmanhlinh) [inline src/error/message-eror.ts]: Use upper case.

File: src/user/dto/CreateUser.dto.ts [hunk 22]
```diff
@@ -15,7 +15,9 @@ export class createUserDTO {
   @MinLength(8)
   password: string;
   @IsNotEmpty()
+
```
Comment (tranmanhlinh) [inline src/user/dto/CreateUser.dto.ts]: Empty lines are between fields, not between decorator and variable.

File: src/user/user.controller.ts [hunk 23]
```diff
@@ -7,29 +7,35 @@ import {
...
 import { createUserDTO } from './dto/CreateUser.dto';
 import { User } from './user.entity';
 import { UserService } from './user.service';
 
 @Controller('user')
 export class UserController {
   constructor(private userService: UserService) {}
+  @UseGuards(AuthGuard())
   @Get('/:id')
+  @UseGuards(AuthGuard())
```
Comment (tranmanhlinh) [inline src/user/user.controller.ts]: Duplicated decorator.

File: src/user/users.repository.ts [hunk 24]
```diff
@@ -18,12 +18,12 @@ export class UserRepository extends Repository<User> {
     const newUser = {
       email,
       password: hashedPassword,
-      first_Name: firstName,
-      last_Name: lastName,
+      firstName: firstName,
+      lastName: lastName,
```
Comment (tranmanhlinh) [inline src/user/users.repository.ts]: { a: a} will simply become { a }

File: src/author/author.service.ts (hunk 1 above)
Comment (mrkiencom) [reply src/author/author.service.ts]: I've tried using AuthorModule inside BookModule, but i can't do that

File: src/author/author.service.ts (hunk 2 above)
Comment (mrkiencom) [reply src/author/author.service.ts]: done

File: src/author/author.service.ts (hunk 3 above)
Comment (mrkiencom) [reply src/author/author.service.ts]: done

File: src/author/author.service.ts (hunk 4 above)
Comment (mrkiencom) [reply src/author/author.service.ts]: done

File: src/error/message-eror.ts (hunk 20 above)
Comment (mrkiencom) [reply src/error/message-eror.ts]: done

File: src/author/author.service.ts (hunk 5 above)
Comment (mrkiencom) [reply src/author/author.service.ts]: done

File: src/error/message-eror.ts (hunk 21 above)
Comment (mrkiencom) [reply src/error/message-eror.ts]: done

File: src/user/dto/CreateUser.dto.ts (hunk 22 above)
Comment (mrkiencom) [reply src/user/dto/CreateUser.dto.ts]: done

File: src/book/book.entity.ts (hunk 7 above)
Comment (mrkiencom) [reply src/book/book.entity.ts:19]: done

File: src/book/book.controller.ts (hunk 6 above)
Comment (mrkiencom) [reply src/book/book.controller.ts]: done

File: src/book/book.entity.ts (hunk 8 above)
Comment (mrkiencom) [reply src/book/book.entity.ts:25]: done

File: src/book/book.repository.ts (hunk 9 above)
Comment (mrkiencom) [reply src/book/book.repository.ts]: done

File: src/book/book.repository.ts (hunk 10 above)
Comment (mrkiencom) [reply src/book/book.repository.ts]: done

Comment (mrkiencom) [reply src/book/book.repository.ts:15]: done

File: src/book/book.service.ts (hunk 13 above)
Comment (mrkiencom) [reply src/book/book.service.ts]: done

File: src/book/book.service.ts (hunk 14 above)
Comment (mrkiencom) [reply src/book/book.service.ts]: done

File: src/book/book.service.ts (hunk 15 above)
Comment (mrkiencom) [reply src/book/book.service.ts]: done

File: src/book/dto/book-dto.ts (hunk 16 above)
Comment (mrkiencom) [reply src/book/dto/book-dto.ts:23]: done

File: src/book/func/get-date.ts (hunk 18 above)
Comment (mrkiencom) [reply src/book/func/get-date.ts:5]: done

File: src/category/category.service.ts (hunk 19 above)
Comment (mrkiencom) [reply src/category/category.service.ts]: done

File: src/author/author.entity.ts [hunk 25]
```diff
@@ -8,4 +9,8 @@ export class Author {
   @Column()
   @IsNotEmpty()
   name: string;
+  @OneToMany((_type) => Book, (book) => book.author, { eager: true })
+  books: Book[];
+  @Column({ name: 'is_deleted', default: false })
+  isDeleted: boolean;
```
Comment (tranmanhlinh) [inline src/author/author.entity.ts:18]: Add empty lines between fields.

File: src/author/author.service.ts [hunk 26]
```diff
@@ -23,41 +25,49 @@ export class AuthorService {
...
+    try {
+      return await this.authorRepository.save({ ...found, isDeleted: true });
+    } catch (error) {
+      throw new InternalServerErrorException(error.message);
     }
   }
+
   async createAuthor(name: string): Promise<Author> {
     const newAuthor = {
       name: name,
```
Comment (tranmanhlinh) [inline src/author/author.service.ts]: `const newAuthor = { name } ` is enough

File: src/book/book.controller.ts [hunk 27]
```diff
@@ -0,0 +1,45 @@
...
+import { Book } from './book.entity';
+import { BookService } from './book.service';
+import { BookDto } from './dto/book-dto';
+import { GetBooksFilterDto } from './dto/book-filter';
+
+@Controller('book')
+@UseGuards(AuthGuard())
+export class BookController {
+  constructor(private bookService: BookService) {}
+  @Get()
```
Comment (tranmanhlinh) [inline src/book/book.controller.ts:22]: add empty line

File: src/book/book.controller.ts [hunk 28]
```diff
@@ -0,0 +1,45 @@
...
+  updateBook(
+    @Param('id') id: string,
+    @Body() newInfoBook: BookDto,
+  ): Promise<Book> {
+    return this.bookService.updateBook(id, newInfoBook);
+  }
+  @Delete('/:id')
+  deleteBook(@Param('id') id: string): Promise<Book> {
+    return this.bookService.deleteBook(id);
+  }
```
Comment (tranmanhlinh) [inline src/book/book.controller.ts:48]: add empty line between methods.

File: src/book/book.service.ts [hunk 29]
```diff
@@ -0,0 +1,159 @@
...
+        categoryId: newCreatedCategory.id,
+        publishYear,
+        price,
+        description,
+        cover,
+        author: newCreatedAuthor,
+        category: newCreatedCategory,
+      });
+    } catch (error) {
+      if (error.code === 23505) {
```
Comment (tranmanhlinh) [inline src/book/book.service.ts]: keep error code in the constants

File: src/author/author.entity.ts (hunk 25 above)
Comment (mrkiencom) [reply src/author/author.entity.ts:18]: done

File: src/author/author.service.ts (hunk 26 above)
Comment (mrkiencom) [reply src/author/author.service.ts]: done

File: src/book/book.service.ts (hunk 29 above)
Comment (mrkiencom) [reply src/book/book.service.ts]: done

File: src/book/book.controller.ts (hunk 28 above)
Comment (mrkiencom) [reply src/book/book.controller.ts:48]: done

**Pull request merged**